### PR TITLE
Feature/idp 823 explicit contract first flow 

### PR DESCRIPTION
**Pull request by PrincessMadMath**
## Description of changes

- Add this required property `DevelopmentMode` to select between ContractFirst and CodeFirst
- Make OasDiff opt-out by default for the ContractFirst mode since required code generation which can cause issue (can be opt-in via `ValidateCodeSync` )
- Support disabling spec generation on the CI for CodeFirst mode: via environement variable `WL_DISABLE_SPECGEN`

## Breaking changes

- Use must now select between ContractFirst and CodeFirst

## Additional checks

- [ ] Updated the documentation of the project to reflect the changes (SOON)
- [X] Added new tests that cover the code changes

### CODE REVIEW

File: src/Workleap.OpenApi.MSBuild/CodeFirstProcess.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+/// For a Code First approach it will:
+///     1. (if not disabled) Generate the OpenAPI specification files from the code
+///     2. Validate the OpenAPI specification files base on spectral rules 
+/// </summary>
+internal class CodeFirstProcess
+{
+    private readonly SpectralManager _spectralManager;
+    private readonly SwaggerManager _swaggerManager;
+
+    internal CodeFirstProcess(SpectralManager spectralManager, SwaggerManager swaggerManager)
```
Comment (asimmon) [inline src/Workleap.OpenApi.MSBuild/CodeFirstProcess.cs:13]: This file looks good overall. Minor comments:

`Environment.GetEnvironmentVariable("WL_DISABLE_SPECGEN") != "true";`
Usually, strings from env vars are compared using ordinal ignore case.

`isGenerationEnable` => `isGenerationEnabled` (+d).

If manually copy pasting the generated spec file from bin/ to the source code dir, handle `IOException` than can occur if the file is already opened and locked by the dev IDE. Print a better error message, followed by the original exception

File: src/Workleap.OpenApi.MSBuild/ContractFirstProcess.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        this._swaggerManager = swaggerManager;
+        this._oasdiffManager = oasdiffManager;
+    }
+
+    internal async Task<bool> Execute(
+        string[] openApiSpecificationFiles,
+        string openApiToolsDirectoryPath,
+        string[] openApiSwaggerDocumentNames,
+        string openApiSpectralRulesetUrl,
+        bool validateCodeSync,
```
Comment (asimmon) [inline src/Workleap.OpenApi.MSBuild/ContractFirstProcess.cs:28]: Not a big fan of booleans as parameters in general. I personally prefer enums.

I don't mind if you want to keep it as is.

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs [hunk 3]
```diff
@@ -4,28 +4,43 @@ namespace Workleap.OpenApi.MSBuild;
...
 {
+    private const string CodeFirst = "CodeFirst";
+    private const string ContractFirst = "ContractFirst";
+    
+    /// <summary>
+    /// 2 supported mode:
+    ///     - CodeFirst: Generate the OpenAPI specification files from the code
+    ///     - ContractFirst: Will use the OpenAPI specification files provided
+    /// </summary>
+    [Microsoft.Build.Framework.Required]
```
Comment (asimmon) [inline src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:15]: Why specifying the full type name here, and the rest of the code isn't like that?

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs [hunk 4]
```diff
@@ -4,28 +4,43 @@ namespace Workleap.OpenApi.MSBuild;
...
+    public string DevelopmentMode { get; set; } = string.Empty;
+
+    /// <summary>When Development mode is Contract first, will validate if the specification match the code.</summary>  
+    [Microsoft.Build.Framework.Required]
+    public bool ValidateCodeSync { get; set; } = false;
```
Comment (asimmon) [inline src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:20]: Please prefix the properties with "OpenApi". It should be clear in the consumer code that a MSBuild property is specific to this tool and not a built-in MSBuild property

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs [hunk 5]
```diff
@@ -4,28 +4,43 @@ namespace Workleap.OpenApi.MSBuild;
 
 public sealed class ValidateOpenApiTask : CancelableAsyncTask
 {
+    private const string CodeFirst = "CodeFirst";
+    private const string ContractFirst = "ContractFirst";
+    
+    /// <summary>
+    /// 2 supported mode:
```
Comment (asimmon) [inline src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:11]: modes (+s).

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs [hunk 6]
```diff
@@ -4,28 +4,43 @@ namespace Workleap.OpenApi.MSBuild;
...
+    /// </summary>
+    [Microsoft.Build.Framework.Required]
+    public string DevelopmentMode { get; set; } = string.Empty;
+
+    /// <summary>When Development mode is Contract first, will validate if the specification match the code.</summary>  
+    [Microsoft.Build.Framework.Required]
+    public bool ValidateCodeSync { get; set; } = false;
+    
     /// <summary>The path of the ASP.NET Core project startup assembly directory.</summary>
     [Required]
```
Comment (asimmon) [inline src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:23]: The full attribute type name isn't specified here

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs [hunk 7]
```diff
@@ -58,23 +76,36 @@ protected override async Task<bool> ExecuteAsync(CancellationToken cancellationT
...
+                case CodeFirst:
+                    await codeFirstProcess.Execute(
+                        this.OpenApiSpecificationFiles,
+                        this.OpenApiSwaggerDocumentNames,
+                        this.OpenApiSpectralRulesetUrl,
+                        cancellationToken);
+                    break;
+                
+                case ContractFirst:
+                    var isSuccess = await contractFirstProcess.Execute(
```
Comment (asimmon) [inline src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:90]: the bool return value serves no purpose, right?

Comment (PrincessMadMath) [reply src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:90]: If we return false it set the task as an error

image

File: src/WebApiDebugger/Controllers/WeatherForecastController.cs [hunk 8]
```diff
@@ -19,7 +19,7 @@ public WeatherForecastController(ILogger<WeatherForecastController> logger)
         this._logger = logger;
     }
 
-    [HttpGet(Name = "GetWeatherForecast")]
+    [HttpGet(Name = "GetWeatherForecast2")]
```
[the remark this answers is not in the document]
Comment (PrincessMadMath) [reply src/WebApiDebugger/Controllers/WeatherForecastController.cs]: Was fixes

File: README.md [hunk 9]
```diff
@@ -1,3 +1,30 @@
 # Workleap.OpenApi.MSBuild
 
 Validates at build time that the OpenAPI specification files extracted from the ASP.NET Core Web API being built conform to Workleap API guidelines.
+
+Depending if the users the Contract-First or Code-First approach this MsBuild task will:
```
Comment (asimmon) [inline README.md:5]:
```suggestion
Depending if the user chose the Contract-First or Code-First development mode this MSBuild task will:
```

File: README.md [hunk 10]
```diff
@@ -1,3 +1,30 @@
...
+
+- Install tools: OasDiff, Spectral, SwashbuckleCLI
```
Comment (asimmon) [inline README.md:7]: Maybe provide links to the tools?

- https://github.com/Tufin/oasdiff
- https://github.com/stoplightio/spectral
- https://github.com/domaindrivendev/Swashbuckle.AspNetCore?tab=readme-ov-file#swashbuckleaspnetcorecli

File: README.md [hunk 11]
```diff
@@ -1,3 +1,30 @@
...
+- Generate the OpenAPI specification file from the associated Web API
+- Validate Spectral rules
```
Comment (asimmon) [inline README.md:9]: Maybe add link? https://github.com/gsoft-inc/wl-api-guidelines/blob/main/.spectral.yaml

File: README.md [hunk 12]
```diff
@@ -1,3 +1,30 @@
...
+- Compare the given OpenAPI specification file with the generated one
+
+## How it work
```
Comment (asimmon) [inline README.md:12]:
```suggestion
## How it works
```

File: README.md [hunk 13]
```diff
@@ -1,3 +1,30 @@
...
+
+- The entry point is `ValidateOpenApiTask.ExecuteAsync()` and will be executed during the build process on project that reference it. This is defined in `./src/Workleap.OpenApi.MSBuild/msbuild/tools/Workleap.OpenApi.MSBuild.targets` as a `UsingTask.TaskName`
```
Comment (asimmon) [inline README.md:14]:
```suggestion
- The entry point is `ValidateOpenApiTask.ExecuteAsync()` and will be executed after the referencing project is built. This is defined in `./src/Workleap.OpenApi.MSBuild/msbuild/tools/Workleap.OpenApi.MSBuild.targets` as a `UsingTask.TaskName`
```

Comment (asimmon) [inline README.md:14]: Maybe link to the official MSBuild docs? https://learn.microsoft.com/en-us/visualstudio/msbuild/tutorial-custom-task-code-generation?view=vs-2022#include-msbuild-properties-and-targets-in-a-package

File: README.md [hunk 14]
```diff
@@ -1,3 +1,30 @@
...
+### How debug the project
+
+Since it's a MSBuild task named `ValidateOpenApi` you can run it this command: `msbuild /t:ValidateOpenApi`. `./src/WebApiDebugger` already have a `launchSettings.json` named `ValidateOpenApi` that you can run in debug.
+
+### With the system test
+
+This command `./Run-SystemTest.ps1` will:
+
+1. Pack the MSBuild library
+2. Build the WebApi.MssBuild.SystemTest and inject the previously packed library
```
Comment (asimmon) [inline README.md:28]:
```suggestion
2. Build the WebApi.MsBuild.SystemTest and inject the previously packed library
```

Sadly we used lowercase 's' in MSBuild...

File: src/Workleap.OpenApi.MSBuild/ContractFirstProcess.cs [hunk 15]
```diff
@@ -0,0 +1,93 @@
+﻿namespace Workleap.OpenApi.MSBuild;
+
+/// <summary>
+/// For a Contract First approach it will:
+///     1. Validate the OpenAPI specification files base on spectral rules
+///     2. If validateCodeSync=true, will generate the OpenAPI specification files from the code and validate if it match the provided specifications.
```
Comment (asimmon) [inline src/Workleap.OpenApi.MSBuild/ContractFirstProcess.cs:6]: `validateCodeSync` n'existe plus. Eventuellement utilise <see cref pour référer à du code et que ça affiche quand un symbole a été retiré

**Review by asimmon — APPROVED** [review]

**Pull request merged**